### DISCUSSION TITLE
Native Apple & Google Sign In w/Capacitor?

### DISCUSSION THREAD

**Discussion by wiverson** [supabase/supabase#9909] [Questions] [answered]
Is there a way to set up native Apple and Google sign in with Supabase using Capacitor?

I have found https://github.com/capacitor-community/apple-sign-in for Apple and https://www.npmjs.com/package/@codetrix-studio/capacitor-google-auth for Google. It looks like this would handle the native integration, but then I need some way to turn the tokens that come back from these into Supabase users?

Any links, documentation, etc. would be very helpful.

**Comment by itsalb3rt** [comment]
You found a solution for this ?

**Comment by wiverson** [comment]
Not really, no.

https://github.com/supabase/supabase-flutter/issues/5
https://github.com/supabase/supabase-js/issues/170

I did manage to get Sign In With Apple working with Capacitor, but not Supabase. I forked the older Capacitor Sign In With Apple and updated it [here](https://github.com/ChangeNode/capacitor-sign-in-with-apple). Here's a demo app showing [how to use it](https://github.com/wiverson/sign-in-with-apple-my-app) w/SvelteKit & Capacitor, tho it's been a few months so I assume it's all out of date now lol.

I was exploring implementing the server-side version of Sign In With Apple and Sign In With Google via Supabase functions in the DB w/PostREST interfaces, and while I could verify the JWTs I wasn't sure how to integrate that with the rest of Supabase GoTrue auth and wound up having to use another system (sigh).

**Reply by itsalb3rt** [reply]
Thanks for the information!

**Comment by matteo6b** [comment]
any update ?

**Reply by itsalb3rt** [reply]
Nope, I use the capacitor plugin https://github.com/CodetrixStudio/CapacitorGoogleAuth

**Comment by matteo6b** [comment]
and how it connect with supabase or other backend for apis protecting with auth?

**Reply by itsalb3rt** [reply]
:/ you need to leave supabase for the auth

**Comment by wiverson** [comment, marked as answer]
Ok, so here's a update - as of 6/4/2023, you can use an API exposed by GoTrue to complete the flow for a native Sign in with Apple.

I wound up using Unity to build my app instead of Capacitor for now (long story). To get this working, I helped check in PRs for the https://github.com/supabase-community/gotrue-csharp repository based on the implementation from the https://github.com/supabase/gotrue-dart and https://github.com/supabase/supabase-flutter/ implementations.

Superficially, it's pretty simple. Here's the flow:

- Create a nonce string (a one-time use string to protect against replay attacks)
- Pass that nonce into the native Sign in with Apple dialog. Assuming success, you'll get a JWT back that includes the nonce. You can unpack that JWT on the client to pull out things like the email address.
- Pass the JWT from Apple along with the nonce hash to Supabase GoTrue. Here is the [CSharp version of this code](https://github.com/supabase-community/gotrue-csharp/blob/master/Gotrue/Api.cs#L205). It takes a provider (apple) the JWT from Apple (the idToken) and the nonce as parameters for the REST call.
- The method will check with Apple's key to verify that the JWT is valid.
- Assuming everything is working, you'll get back a session from Supabase and now you are off to the races.

FWIW I did publish an updated version of a Capacitor 4 plugin for the the most current version of the native Sign in with Apple dialog. https://github.com/ChangeNode/capacitor-sign-in-with-apple

I haven't looked at the latest and greatest version of the Supabase JS client to see if they added the method for sign in. But, FWIW it should be very easy to add (it's just a single method).

The only tricky part is that the nonce and the hash need to be a very specific algo or you can run into trouble. A C# version of these can be found in https://github.com/supabase-community/gotrue-csharp/blob/master/Gotrue/Helpers.cs - they should be pretty easy to crib into other languages e.g. JS/TS.

**Reply by matteo6b** [reply]
i can start the explication in JavaScript  with plugin https://www.npmjs.com/package/@codetrix-studio/capacitor-google-auth

the library  in supabase-js don't find the function but in supabase/gotrue-js  it is exits https://github.com/search?q=repo%3Asupabase%2Fgotrue-js%20%20signInWithIdToken&type=code

Throw me a error

Passed nonce and nonce in id_token should either both exist or not





 i don't pass any nonce string because i  don't  know if is a option with plugin capacitor-google-auth

**Reply by wiverson** [reply]
So, I think the idea (at least in the Unity versions of sign in with Google) is that you are basically following the OAuth flow locally. AFAIK the Google Play native dialog stuff is kind of in a weird not supported state, and so the way it more or less has to work for native is via an OAuth flow. Basically you bounce the user to the local browser and then include a deep link back to open up your app.

Here's the best option I found for Unity: https://github.com/hippogamesunity/SimpleGoogleSignIn/wiki

Here's the code and info https://assetstore.unity.com/packages/tools/integration/simple-google-sign-in-250663

In theory there's no reason you can't follow these steps and redo the code in JS. Ahem. O.o

I think there should be a way to essentially recreate these steps but just use the standard Google flow from Supabase. Haven't tested any of this stuff, as I found this stuff easier to get working via Unity.
